### PR TITLE
Feature/issue 70

### LINKED ISSUE #70: Additional Quickfix Suggestions
Some errors are hard to figure out on your own and it would be helpful to have suggestions/quickfixes. See below for a list. Additions are appreciated.

* String literals with " " brackets instead of ' '  -> Quickfix replaces type of brackets
* Numbers in feature names without " " brackets --> Quickfix adds brackets
* Language constructs of not included language levels --> Quickfix adds the language level
* Missing indendation for lower feature -> Quickfix adds indent

### PR DESCRIPTION
**Pull request by felixrieg**
Fixes #70 

As stated in the issue, this adds quickactions for:

-     String literals with " " brackets instead of ' ' -> Quickfix replaces type of brackets
-     Numbers in feature names without " " brackets --> Quickfix adds brackets
-     Language constructs of not included language levels --> Quickfix adds the language level
-     Missing indendation for lower feature -> Quickfix adds indent

### CODE REVIEW

**Review by TobiUUlm — COMMENTED** [review]
Code LGTM,
- Feature name starting with number
- unresolved reference => add '..'
don't work for me.
Adding include and indentation seems to work fine.

File: uvls/src/core/actions.rs [hunk 1]
```diff
@@ -60,3 +61,265 @@ pub fn rename_dash(
...
+        };
+        return Ok(Some(vec![CodeActionOrCommand::CodeAction(
+            code_action_replace,
+        )]));
+    } else {
+        return Ok(None);
+    }
+}
+
+pub fn starts_with_number(
```
[the remark this answers is not in the document]
Comment (felixrieg) [reply uvls/src/core/actions.rs:180]: I fixed this, pls try again

File: uvls/src/core/actions.rs [hunk 2]
```diff
@@ -60,3 +61,265 @@ pub fn rename_dash(
         return Ok(None);
     }
 }
+
+pub fn reference_to_string(
```
[the remark this answers is not in the document]
Comment (felixrieg) [reply uvls/src/core/actions.rs:65]: I don't think this was part of this issue.

File: uvls/src/core/check.rs [hunk 3]
```diff
@@ -253,6 +253,35 @@ pub fn classify_error(root: Node, source: &Rope) -> ErrorInfo {
             };
         }
     }
+    if Regex::new(r"^\d+")
```
Comment (TobiUUlm) [inline uvls/src/core/check.rs:256]: Now the quickfix is available, but it results in a fix that treats the attribute declaration as part of the feature name.  The regex (or whatever recognizes the name) should only match up until the firrst space or closing-quote appears.

2023-10-10 15_49_51- Extension Development Host  ● imtest uvl - uvl - Visual Studio Code   

quickfixes result in: 
`Feature {attrtest 1}_3` or `"3Feature {attrtest 1}"`

Comment (felixrieg) [reply uvls/src/core/check.rs:256]: Fixed it. Pls try again

**Comment by TobiUUlm** [comment]
@felixrieg The number-fix looks fine now.  
unresolved reference => add '..' still did not work.

**Comment by felixrieg** [comment]
@TobiUUlm I added now, that `"one" == two` would get the quickfix to change to `'one' == two`. But when two has a "unresolved reference" there is no quickfix, since this is not part of the issue

**Comment by MartinMUU** [comment]
The new line bug is solved.

There is now a strange behavior when I write e.g. 2Realtype then the quickfix generates some weird things.  But this is only with explicit types the case.

grafik

The autocompleting also works now. Maybe my laptop had some problems during the last revision.

**Comment by felixrieg** [comment]
@MartinMUU I fixed it. And thank you for all the bug search.

**Comment by MartinMUU** [comment]
I still get the poblem. It seems like there are only 6 letters converted after a type. without a type there is no limit of the letters.
grafik

Sometimes i also get the new line bug. But that happens not that often and i cant find a pattern when it occurs.
grafik

**Comment by MartinMUU** [comment]
In general it works now. But sometimes i get weired bugs. E.g. sometimes the quotes or the underscore are just added in front of the word.
grafik

Also sometimes the new line bug is happening. But i can't find a pattern. it feels really random if these bugs occurs.
grafik

**Comment by felixrieg** [comment]
@MartinMUU I can't reproduce this bug...

**Review by MartinMUU — APPROVED** [review]
Seems to work fine now

**Comment by SundermannC** [comment]
Almost everything worked great however I am not sure how to trigger the quickfix for the 4th aspect (Missing indendation for lower feature -> Quickfix adds indent). When can this be applied? 

Also, there seems to be an issue with type constraints violating the includes. Here, no quickfix is shown. Might be related to #121

**Comment by felixrieg** [comment]
> Almost everything worked great however I am not sure how to trigger the quickfix for the 4th aspect (Missing indendation for lower feature -> Quickfix adds indent). When can this be applied?

This quick fix can be used when having a Feature on Root level: 
```
features
A
```
A quick fix is always corresponding to a error and thus we need an error first to provide a quick fix. If you have another error that should provide the same quick fix, we can add it easily

> Also, there seems to be an issue with type constraints violating the includes. Here, no quickfix is shown. Might be related to https://github.com/Universal-Variability-Language/uvl-lsp/issues/121

The Issue described in https://github.com/Universal-Variability-Language/uvl-lsp/issues/121 is because the lsp sometimes stops looking for errors when it already found one that is prior in the analysis. I think this might be a hard thing to fix.
For the given example, I added the quick fix to this error Message.

**Comment by SundermannC** [comment]
The fix for the string constraints worked great for me now. Thanks a lot!

Ah ok, I tested it only after group keywords like in this example:

image
The error message is more generic here though. It might cause issues to add the quickfix there.

**Pull request merged**